File: handlers/pjsk/assets.go
package pjsk

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

const assetsCacheDir = "cache/pjsk"

var allServers = []string{"jp", "cn", "en", "tw", "kr"}

// commitSHAs 内存中保存的各服务器最新 commit SHA
var commitSHAs sync.Map // server -> string

func repoOwner() string { return "Sekai-World" }

// serverToRepoCode 服务器代码到仓库代码的映射（tw -> tc）
var serverToRepoCode = map[string]string{
	"tw": "tc",
}

func dbDiffName(server string) string {
	if server == "jp" {
		return "sekai-master-db-diff"
	}
	code := server
	if mapped, ok := serverToRepoCode[server]; ok {
		code = mapped
	}
	return "sekai-master-db-" + code + "-diff"
}

func serverCacheDir(server string) string {
	return filepath.Join(assetsCacheDir, dbDiffName(server))
}

func remoteURL(server, file string) string {
	return "https://sekai-world.github.io/" + dbDiffName(server) + "/" + file
}


// ghContentsEntry GitHub Contents API 返回的单个条目
type ghContentsEntry struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

// ghCommit GitHub Commits API 返回的条目
type ghCommit struct {
	SHA string `json:"sha"`
}

const commitSHAFile = ".commit_sha"

// githubToken 从环境变量读取 GitHub API Token
var githubToken = os.Getenv("GITHUB_TOKEN")

// newGitHubRequest 创建带认证的 GitHub API 请求
func newGitHubRequest(url string) (*http.Request, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "amiabot-pages/1.0")
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	if githubToken != "" {
		req.Header.Set("Authorization", "Bearer "+githubToken)
	}
	return req, nil
}

// fetchLatestCommitSHA 通过 GitHub API 获取仓库 main 分支最新 commit SHA
func fetchLatestCommitSHA(server string) (string, error) {
	repo := dbDiffName(server)
	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/commits?sha=main&per_page=1", repoOwner(), repo)

	req, err := newGitHubRequest(url)
	if err != nil {
		return "", err
	}

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("请求 GitHub API 失败 (%s): %w", server, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("GitHub API 返回 %d (%s)", resp.StatusCode, server)
	}

	var commits []ghCommit
	if err := json.NewDecoder(resp.Body).Decode(&commits); err != nil {
		return "", fmt.Errorf("解析 commit 响应失败 (%s): %w", server, err)
	}
	if len(commits) == 0 {
		return "", fmt.Errorf("未获取到 commit (%s)", server)
	}
	return commits[0].SHA, nil
}

// loadSavedSHA 从文件加载已保存的 commit SHA
func loadSavedSHA(server string) string {
	path := filepath.Join(serverCacheDir(server), commitSHAFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

// saveSHA 保存 commit SHA 到文件和内存
func saveSHA(server, sha string) {
	commitSHAs.Store(server, sha)
	dir := serverCacheDir(server)
	_ = os.MkdirAll(dir, 0o755)
	_ = os.WriteFile(filepath.Join(dir, commitSHAFile), []byte(sha), 0o644)
}

// fetchFileList 通过 GitHub Contents API 获取仓库根目录所有 .json 文件名
func fetchFileList(server string) ([]string, error) {
	repo := dbDiffName(server)
	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/contents/", repoOwner(), repo)

	req, err := newGitHubRequest(url)
	if err != nil {
		return nil, err
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("请求 GitHub API 失败 (%s): %w", server, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GitHub API 返回 %d (%s)", resp.StatusCode, server)
	}

	var entries []ghContentsEntry
	if err := json.NewDecoder(resp.Body).Decode(&entries); err != nil {
		return nil, fmt.Errorf("解析 GitHub API 响应失败 (%s): %w", server, err)
	}

	var files []string
	for _, e := range entries {
		if e.Type == "file" && strings.HasSuffix(e.Name, ".json") {
			files = append(files, e.Name)
		}
	}
	return files, nil
}

func downloadFile(server, file string) error {
	url := remoteURL(server, file)
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d for %s", resp.StatusCode, url)
	}

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	dir := serverCacheDir(server)
	_ = os.MkdirAll(dir, 0o755)
	return os.WriteFile(filepath.Join(dir, file), data, 0o644)
}

// refreshServer 刷新单个服务器：检查 commit SHA 和文件完整性，无变化且完整则跳过
func refreshServer(server string, maxConcurrency int, force bool) map[string]string {
	// 获取远程最新 commit SHA
	remoteSHA, err := fetchLatestCommitSHA(server)
	if err != nil {
		return map[string]string{"_error": err.Error()}
	}

	files, err := fetchFileList(server)
	if err != nil {
		return map[string]string{"_error": err.Error()}
	}

	if !force {
		// 对比本地 SHA
		localSHA := ""
		if v, ok := commitSHAs.Load(server); ok {
			localSHA = v.(string)
		} else {
			localSHA = loadSavedSHA(server)
		}

		if localSHA == remoteSHA {
			// SHA 相同，检查本地文件完整性
			missing := 0
			dir := serverCacheDir(server)
			for _, f := range files {
				if _, err := os.Stat(filepath.Join(dir, f)); err != nil {
					missing++
				}
			}
			if missing == 0 {
				return map[string]string{"_skipped": "commit SHA 未变化且文件完整: " + remoteSHA[:12]}
			}
			log.Printf("[pjsk] %s: SHA 未变化但缺少 %d 个文件，继续下载", server, missing)
		}
	}

	type result struct {
		file string
		err  error
	}

	ch := make(chan result, len(files))
	sem := make(chan struct{}, maxConcurrency)
	var wg sync.WaitGroup

	for _, f := range files {
		wg.Add(1)
		go func(file string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			var dlErr error
			for attempt := 0; attempt < 3; attempt++ {
				if dlErr = downloadFile(server, file); dlErr == nil {
					break
				}
				time.Sleep(time.Duration(attempt+1) * time.Second)
			}
			ch <- result{file: file, err: dlErr}
		}(f)
	}

	wg.Wait()
	close(ch)

	results := make(map[string]string, len(files))
	for r := range ch {
		if r.err != nil {
			results[r.file] = r.err.Error()
		} else {
			results[r.file] = "ok"
		}
	}

	// 下载完成后保存新的 commit SHA
	saveSHA(server, remoteSHA)
	results["_commit"] = remoteSHA

	return results
}

// RefreshAll 并发刷新所有服务器（每个服务器内部也并发下载）
func RefreshAll(force bool) map[string]map[string]string {
	var mu sync.Mutex
	allResults := make(map[string]map[string]string, len(allServers))
	var wg sync.WaitGroup

	for _, s := range allServers {
		wg.Add(1)
		go func(server string) {
			defer wg.Done()
			r := refreshServer(server, 20, force)
			mu.Lock()
			allResults[server] = r
			mu.Unlock()
		}(s)
	}

	wg.Wait()
	return allResults
}

// ReadCachedJSON 读取缓存的 JSON 文件
func ReadCachedJSON(server, file string) ([]byte, error) {
	path := filepath.Join(serverCacheDir(server), file)
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("缓存文件不存在: %s/%s，请先调用 /pjsk/assets/refresh", server, file)
	}
	return data, nil
}

// InitAssets 启动时加载 commit SHA，后台全量刷新
func InitAssets() {
	for _, s := range allServers {
		if sha := loadSavedSHA(s); sha != "" {
			commitSHAs.Store(s, sha)
		}
	}

	log.Println("[pjsk] 后台启动全量下载...")
	go func() {
		results := RefreshAll(false)
		total, failed := 0, 0
		for server, serverResults := range results {
			for key, status := range serverResults {
				if strings.HasPrefix(key, "_") {
					log.Printf("[pjsk] %s: %s = %s", server, key, status)
					continue
				}
				total++
				if status != "ok" {
					failed++
					log.Printf("[pjsk] 下载失败: %s 原因: %s", remoteURL(server, key), status)
				}
			}
		}
		log.Printf("[pjsk] 全量下载完成: %d 个文件, %d 个失败", total, failed)
	}()
}

// AssetHandler GET /pjsk/assets/*path
// /pjsk/assets/refresh → 刷新缓存
// /pjsk/assets/sekai-master-db-diff/xxx.json → 返回缓存 JSON
func AssetHandler(c *gin.Context) {
	raw := strings.TrimPrefix(c.Param("path"), "/")

	if raw == "refresh" {
		force := c.Query("force") == "true"
		results := RefreshAll(force)
		c.JSON(http.StatusOK, gin.H{"results": results})
		return
	}

	parts := strings.SplitN(raw, "/", 2)
	if len(parts) != 2 || parts[1] == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效路径"})
		return
	}

	dirName, file := parts[0], parts[1]

	var server string
	if dirName == "sekai-master-db-diff" {
		server = "jp"
	} else if strings.HasPrefix(dirName, "sekai-master-db-") && strings.HasSuffix(dirName, "-diff") {
		server = strings.TrimSuffix(strings.TrimPrefix(dirName, "sekai-master-db-"), "-diff")
	} else {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的仓库名: " + dirName})
		return
	}

	if !validServers[server] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "无效的服务器: " + server})
		return
	}

	data, err := ReadCachedJSON(server, file)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
		return
	}

	c.Data(http.StatusOK, "application/json; charset=utf-8", data)
}

File: src/handlers/pjsk/assets.rs
use crate::pkg::http_error::format_upstream_http_error;
use axum::{
    extract::{Path as AxumPath, Query},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use once_cell::sync::Lazy;
use serde::Deserialize;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::{RwLock, Semaphore};

use crate::handlers::pjsk::VALID_SERVERS;

const MASTER_DATA_CACHE_DIR: &str = "cache/pjsk";
const COMMIT_SHA_FILE: &str = ".commit_sha";

pub static COMMIT_SHAS: Lazy<Arc<RwLock<HashMap<String, String>>>> =
    Lazy::new(|| Arc::new(RwLock::new(HashMap::new())));

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
pub struct RefreshQuery {
    pub force: Option<String>,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct GhCommit {
    sha: String,
}

#[derive(Deserialize, Debug)]
#[allow(non_snake_case)]
#[allow(dead_code)]
struct GhContentsEntry {
    name: String,
    r#type: String,
}

fn server_to_repo_code(server: &str) -> &str {
    match server {
        "tw" => "tc",
        other => other,
    }
}

pub fn db_diff_name(server: &str) -> String {
    if server == "jp" {
        "sekai-master-db-diff".to_string()
    } else {
        format!("sekai-master-db-{}-diff", server_to_repo_code(server))
    }
}

pub fn server_cache_dir(server: &str) -> PathBuf {
    Path::new(MASTER_DATA_CACHE_DIR).join(db_diff_name(server))
}

fn remote_url(server: &str, file: &str) -> String {
    format!(
        "https://sekai-world.github.io/{}/{}",
        db_diff_name(server),
        file
    )
}

fn github_request(url: &str) -> Result<reqwest::RequestBuilder, String> {
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(30))
        .build()
        .map_err(|e| e.to_string())?;

    let mut builder = client
        .get(url)
        .header("User-Agent", "amiabot-pages/1.0")
        .header("Accept", "application/vnd.github.v3+json");

    if let Ok(token) = env::var("GITHUB_TOKEN") {
        if !token.trim().is_empty() {
            builder = builder.header("Authorization", format!("Bearer {}", token.trim()));
        }
    }
    Ok(builder)
}

async fn fetch_latest_commit_sha(server: &str) -> Result<String, String> {
    let repo = db_diff_name(server);
    let url = format!(
        "https://api.github.com/repos/Sekai-World/{}/commits?sha=main&per_page=1",
        repo
    );

    let builder = github_request(&url)?;
    let resp = crate::pkg::http_client::send(builder)
        .await
        .map_err(|e| format!("请求 GitHub API 失败 ({}): {}", server, e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format_upstream_http_error(
            &format!("GitHub API ({server})"),
            status,
            &body,
        ));
    }

    let commits = resp
        .json::<Vec<GhCommit>>()
        .await
        .map_err(|e| format!("解析 commit 响应失败 ({}): {}", server, e))?;
    if commits.is_empty() {
        return Err(format!("未获取到 commit ({})", server));
    }
    Ok(commits[0].sha.clone())
}

fn load_saved_sha(server: &str) -> String {
    let path = server_cache_dir(server).join(COMMIT_SHA_FILE);
    fs::read_to_string(path)
        .map(|s| s.trim().to_string())
        .unwrap_or_default()
}

async fn save_sha(server: &str, sha: &str) {
    let mut shas = COMMIT_SHAS.write().await;
    shas.insert(server.to_string(), sha.to_string());
    drop(shas);

    let dir = server_cache_dir(server);
    let _ = fs::create_dir_all(&dir);
    let _ = fs::write(dir.join(COMMIT_SHA_FILE), sha.as_bytes());
}

async fn fetch_file_list(server: &str) -> Result<Vec<String>, String> {
    let repo = db_diff_name(server);
    let url = format!(
        "https://api.github.com/repos/Sekai-World/{}/contents/",
        repo
    );

    let builder = github_request(&url)?;
    let resp = crate::pkg::http_client::send(builder)
        .await
        .map_err(|e| format!("请求 GitHub API 失败 ({}): {}", server, e))?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format_upstream_http_error(
            &format!("GitHub API ({server})"),
            status,
            &body,
        ));
    }

    let entries = resp
        .json::<Vec<GhContentsEntry>>()
        .await
        .map_err(|e| format!("解析 GitHub API 响应失败 ({}): {}", server, e))?;
    let mut files = Vec::new();
    for entry in entries {
        if entry.r#type == "file" && entry.name.ends_with(".json") {
            files.push(entry.name);
        }
    }
    Ok(files)
}

async fn download_file(server: &str, file: &str) -> Result<(), String> {
    let url = remote_url(server, file);
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(60))
        .build()
        .unwrap_or_default();

    let resp = crate::pkg::http_client::send(
        client
            .get(&url)
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"),
    )
    .await
    .map_err(|e| e.to_string())?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format_upstream_http_error(
            &format!("masterdata 下载 {url}"),
            status,
            &body,
        ));
    }

    let bytes = resp.bytes().await.map_err(|e| e.to_string())?;
    let path = server_cache_dir(server).join(file);
    let _ = fs::create_dir_all(path.parent().unwrap());
    fs::write(path, bytes).map_err(|e| e.to_string())?;
    Ok(())
}

pub async fn refresh_server(
    server: &str,
    max_concurrency: usize,
    force: bool,
) -> HashMap<String, String> {
    let mut results = HashMap::new();

    let remote_sha = match fetch_latest_commit_sha(server).await {
        Ok(sha) => sha,
        Err(e) => {
            results.insert("_error".to_string(), e);
            return results;
        }
    };

    let files = match fetch_file_list(server).await {
        Ok(f) => f,
        Err(e) => {
            results.insert("_error".to_string(), e);
            return results;
        }
    };

    if !force {
        let shas = COMMIT_SHAS.read().await;
        let local_sha = shas
            .get(server)
            .cloned()
            .unwrap_or_else(|| load_saved_sha(server));
        drop(shas);

        if local_sha == remote_sha {
            let mut missing = 0;
            let dir = server_cache_dir(server);
            for file in &files {
                if !dir.join(file).exists() {
                    missing += 1;
                }
            }
            if missing == 0 {
                results.insert(
                    "_skipped".to_string(),
                    format!(
                        "commit SHA 未变化且文件完整: {}",
                        &remote_sha[..12.min(remote_sha.len())]
                    ),
                );
                return results;
            }
            tracing::info!(
                %server,
                missing,
                "masterdata SHA 未变化但缺少文件，继续下载"
            );
        }
    }

    let sem = Arc::new(Semaphore::new(max_concurrency));
    let mut tasks = Vec::new();

    for file in files.clone() {
        let sem_clone = Arc::clone(&sem);
        let s = server.to_string();
        tasks.push(tokio::spawn(async move {
            let _permit = sem_clone.acquire().await.unwrap();
            let mut dl_err = None;
            for attempt in 0..3 {
                match download_file(&s, &file).await {
                    Ok(_) => {
                        dl_err = None;
                        break;
                    }
                    Err(e) => {
                        dl_err = Some(e);
                        tokio::time::sleep(Duration::from_secs(attempt + 1)).await;
                    }
                }
            }
            (file, dl_err)
        }));
    }

    let mut has_error = false;
    for task in tasks {
        if let Ok((file, err)) = task.await {
            if let Some(e) = err {
                results.insert(file.clone(), e);
                has_error = true;
            } else {
                results.insert(file.clone(), "ok".to_string());
            }
        }
    }

    if has_error {
        results.insert(
            "_commit".to_string(),
            "not_saved (存在下载失败的文件)".to_string(),
        );
    } else {
        save_sha(server, &remote_sha).await;
        results.insert("_commit".to_string(), remote_sha);
    }

    results
}

pub async fn refresh_all(force: bool) -> HashMap<String, HashMap<String, String>> {
    let servers = vec!["jp", "cn", "en", "tw", "kr"];
    let mut all_results = HashMap::new();
    let mut tasks = Vec::new();

    for s in servers {
        tasks.push(tokio::spawn(async move {
            let r = refresh_server(s, 20, force).await;
            (s.to_string(), r)
        }));
    }

    for task in tasks {
        if let Ok((s, r)) = task.await {
            all_results.insert(s, r);
        }
    }

    all_results
}

pub fn read_cached_json(server: &str, file: &str) -> Result<Vec<u8>, String> {
    let path = server_cache_dir(server).join(file);
    fs::read_to_string(path)
        .map(|s| s.into_bytes())
        .map_err(|_| format!("缓存文件不存在: {}/{}，请先刷新", server, file))
}

pub async fn init_master_data() {
    let servers = vec!["jp", "cn", "en", "tw", "kr"];
    let mut shas = COMMIT_SHAS.write().await;
    for s in servers {
        let sha = load_saved_sha(s);
        if !sha.is_empty() {
            shas.insert(s.to_string(), sha);
        }
    }
    drop(shas);

    tracing::info!("开始加载 masterdata（等待完成后服务才会启动）");
    let results = refresh_all(false).await;
    let (mut total, mut failed) = (0, 0);
    for (server, server_results) in results {
        for (key, status) in server_results {
            if key.starts_with('_') {
                tracing::info!(%server, %key, %status, "masterdata 状态");
                continue;
            }
            total += 1;
            if status != "ok" {
                failed += 1;
                tracing::warn!(
                    url = %remote_url(&server, &key),
                    error = %status,
                    "masterdata 文件下载失败"
                );
            }
        }
    }
    tracing::info!(total, failed, "masterdata 加载完成");
}

pub async fn master_data_handler(
    AxumPath(raw_path): AxumPath<String>,
    Query(q): Query<RefreshQuery>,
) -> impl IntoResponse {
    let raw = raw_path.trim_start_matches('/');

    if raw == "refresh" {
        let force = q.force.map(|s| s == "true").unwrap_or(false);
        let results = refresh_all(force).await;
        return (StatusCode::OK, Json(serde_json::to_value(results).unwrap())).into_response();
    }

    let parts: Vec<&str> = raw.splitn(2, '/').collect();
    if parts.len() != 2 || parts[1].is_empty() {
        return (StatusCode::BAD_REQUEST, "无效路径").into_response();
    }

    let dir_name = parts[0];
    let file = parts[1];

    let server = if dir_name == "sekai-master-db-diff" {
        "jp"
    } else if dir_name.starts_with("sekai-master-db-") && dir_name.ends_with("-diff") {
        &dir_name["sekai-master-db-".len()..dir_name.len() - "-diff".len()]
    } else {
        return (
            StatusCode::BAD_REQUEST,
            format!("无效的仓库名: {}", dir_name),
        )
            .into_response();
    };

    if !VALID_SERVERS.contains(server) {
        return (StatusCode::BAD_REQUEST, format!("无效的服务器: {}", server)).into_response();
    }

    match read_cached_json(server, file) {
        Ok(data) => (
            StatusCode::OK,
            [(
                axum::http::header::CONTENT_TYPE,
                "application/json; charset=utf-8",
            )],
            data,
        )
            .into_response(),
        Err(e) => (StatusCode::NOT_FOUND, e).into_response(),
    }
}
